cartographer_ros/cartographer_ros/msg_conversion.cc

```diff
@@ -183,29 +219,85 @@ ToPointCloudWithIntensities(const sensor_msgs::MultiEchoLaserScan& msg) {
 
 std::tuple<::cartographer::sensor::PointCloudWithIntensities,
            ::cartographer::common::Time>
-ToPointCloudWithIntensities(const sensor_msgs::PointCloud2& message) {
+ToPointCloudWithIntensities(const sensor_msgs::PointCloud2& msg) {
   PointCloudWithIntensities point_cloud;
   // We check for intensity field here to avoid run-time warnings if we pass in
   // a PointCloud2 without intensity.
-  if (PointCloud2HasField(message, "intensity")) {
-    pcl::PointCloud<pcl::PointXYZI> pcl_point_cloud;
-    pcl::fromROSMsg(message, pcl_point_cloud);
+  if (PointCloud2HasField(msg, "intensity")) {
+    if (PointCloud2HasField(msg, "time")) {
+      pcl::PointCloud<PointXYZIT> pcl_point_cloud;
+      pcl::fromROSMsg(msg, pcl_point_cloud);
+      for (const auto& point : pcl_point_cloud) {
+        point_cloud.points.emplace_back(point.x, point.y, point.z, point.time);
+        point_cloud.intensities.push_back(point.intensity);
+      }
+    } else {
+      pcl::PointCloud<pcl::PointXYZI> pcl_point_cloud;
+      pcl::fromROSMsg(msg, pcl_point_cloud);
+      for (const auto& point : pcl_point_cloud) {
+        point_cloud.points.emplace_back(point.x, point.y, point.z, 0.f);
+        point_cloud.intensities.push_back(point.intensity);
+      }
+    }
+  } else {
+    // If we don't have an intensity field, just copy XYZ and fill in 1.0.
+    if (PointCloud2HasField(msg, "time")) {
+      pcl::PointCloud<PointXYZT> pcl_point_cloud;
+      pcl::fromROSMsg(msg, pcl_point_cloud);
+      for (const auto& point : pcl_point_cloud) {
+        point_cloud.points.emplace_back(point.x, point.y, point.z, point.time);
+        point_cloud.intensities.push_back(1.0);
+      }
+    } else {
+      pcl::PointCloud<pcl::PointXYZ> pcl_point_cloud;
+      pcl::fromROSMsg(msg, pcl_point_cloud);
+      for (const auto& point : pcl_point_cloud) {
+        point_cloud.points.emplace_back(point.x, point.y, point.z, 0.f);
+        point_cloud.intensities.push_back(1.0);
+      }
+    }
+  }
+  ::cartographer::common::Time timestamp = FromRos(msg.header.stamp);
+  if (!point_cloud.points.empty()) {
+    const double duration = point_cloud.points.back()[3];
+    timestamp += cartographer::common::FromSeconds(duration);
+    for (Eigen::Vector4f& point : point_cloud.points) {
+      point[3] -= duration;
+      CHECK_LE(point[3], 0) << "Encountered a point with a larger stamp than "
+                               "the last point in the cloud.";
+    }
+  }
+  return std::make_tuple(point_cloud, timestamp);
+}
+
+std::tuple<::cartographer::sensor::TimedPointCloud,
+           ::cartographer::common::Time>
+ToTimedPointCloud(const sensor_msgs::PointCloud2& msg) {
```

Comment:
Drop this function. The only caller is SensorBridge::HandlePointCloud2Message and can use ToPointCloudWithIntensities instead.